app/code/Magento/CatalogImportExport/Model/Import/Product.php

```diff
@@ -2972,4 +2991,64 @@ private function getExistingSku($sku)
     {
         return $this->_oldSku[strtolower($sku)];
     }
+
+    /**
+     * Set product images 'disable' = 0 for specified store.
+     *
+     * @param array $mediaGalleryData
+     * @return array
+     */
+    private function restoreDisableImage(array $mediaGalleryData)
+    {
+        $restoreData = [];
+        foreach (array_keys($mediaGalleryData) as $storeId) {
+            foreach ($mediaGalleryData[$storeId] as $productSku => $mediaGalleryRows) {
+                $productId = $this->skuProcessor->getNewSku($productSku)[$this->getProductEntityLinkField()];
+                $restoreData[] = sprintf(
+                    'store_id = %s and %s = %s',
+                    $storeId,
+                    $this->getProductEntityLinkField(),
+                    $productId
+                );
+                if (isset($mediaGalleryRows['all']['restore'])) {
+                    unset($mediaGalleryData[$storeId][$productSku]);
+                }
+            }
+        }
+
+        $this->_connection->update(
+            $this->mediaGalleryValueTableName,
+            ['disabled' => 0],
+            new \Zend_Db_Expr(implode(' or ', $restoreData))
+        );
+
+        return $mediaGalleryData;
+    }
+
+    /**
+     * Remove existed images from insert data.
+     *
+     * @param array $multiInsertData
+     * @param array $imageNames
+     * @return array
+     */
+    private function filterImageInsertData(array $multiInsertData, array $imageNames)
+    {
+        //Remove image duplicates for stores.
+        $multiInsertData = array_map("unserialize", array_unique(array_map("serialize", $multiInsertData)));
```

Comment:
Is this format json? I would be a bit worried that we are calling unserialize/serialize on data that we do not trust/have control over.

Reply:
Removed.